File: apps/designer/src/core/scene/sceneDirector.ts
import { DEFAULT_DIALOGUE_WIDTH_PERCENT, DEFAULT_PROP_Z } from "./actions";
import { AssetReferenceCounter, assetReferenceKey } from "./assetReferences";
import {
  offstagePosition,
  resolvePosition,
  STAGE_HEIGHT,
  STAGE_WIDTH,
  type Direction,
  type StagePosition,
  type StageVector,
} from "./positions";
import type { SceneOp } from "./sceneOps";
import { lerp, rafScheduler, runTween, type Easing, type FrameScheduler } from "./transitions";

export type ScenePropState = {
  id: string;
  assetId: string;
  variationId?: string;
  /** Reference key for the currently rendered asset. */
  assetKey: string;
  /** Authored destination: named slot or explicit coordinates. */
  position: StagePosition;
  /** Live stage coordinates, which a transition may be interpolating. */
  x: number;
  y: number;
  visible: boolean;
  opacity: number;
  zIndex: number;
  scale: number;
  highlighted: boolean;
};

export type SceneBackgroundState = {
  assetId: string;
  assetKey: string;
  opacity: number;
  /** Slide offset in stage units; 0,0 is fully on screen. */
  offsetX: number;
  offsetY: number;
};

export type SceneSnapshot = {
  background: SceneBackgroundState | null;
  /** Retained only while a background transition is running. */
  outgoingBackground: SceneBackgroundState | null;
  /** Ordered back to front. */
  props: ScenePropState[];
  /** When false, the dialogue box is hidden even if it has text. Defaults true. */
  dialogueVisible: boolean;
  /** Optional actor id for later per-character dialogue theming. */
  dialogueCharacterId?: string;
  /** Dialogue box width as a percent of stage width. */
  dialogueWidthPercent: number;
  loadedAssetKeys: string[];
};

export type SceneDirectorOptions = {
  scheduler?: FrameScheduler;
  /** Complete every transition immediately; used by previews and tests. */
  skipTransitions?: boolean;
  easing?: Easing;
  onLoadAsset?: (key: string) => void;
  onUnloadAsset?: (key: string) => void;
};

export class SceneActionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SceneActionError";
  }
}

function missingProp(id: string, verb: string): SceneActionError {
  return new SceneActionError(
    `Cannot ${verb} prop '${id}': no active prop with this ID exists. Call Prop.Add(...) before attempting to ${verb} it.`
  );
}

/**
 * Owns the authoritative scene state.
 *
 * Every visual action resolves against this state rather than poking renderer
 * objects directly, so transitions always know where they are starting from.
 */
export class SceneDirector {
  private background: SceneBackgroundState | null = null;
  private outgoing: SceneBackgroundState | null = null;
  private readonly props = new Map<string, ScenePropState>();
  private dialogueVisible = true;
  private dialogueCharacterId: string | undefined;
  private dialogueWidthPercent = DEFAULT_DIALOGUE_WIDTH_PERCENT;
  private backgroundTweenAbort: AbortController | null = null;
  private readonly assets: AssetReferenceCounter;
  private readonly listeners = new Set<() => void>();
  private readonly scheduler: FrameScheduler;
  private readonly skipTransitions: boolean;
  private readonly easing: Easing;
  private cachedSnapshot: SceneSnapshot | null = null;

  constructor(options: SceneDirectorOptions = {}) {
    this.scheduler = options.scheduler ?? rafScheduler;
    this.skipTransitions = options.skipTransitions ?? false;
    this.easing = options.easing ?? "easeInOut";
    this.assets = new AssetReferenceCounter({
      onLoad: options.onLoadAsset,
      onUnload: options.onUnloadAsset,
    });
  }

  subscribe(listener: () => void): () => void {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }

  getSnapshot(): SceneSnapshot {
    if (!this.cachedSnapshot) {
      this.cachedSnapshot = {
        background: this.background ? { ...this.background } : null,
        outgoingBackground: this.outgoing ? { ...this.outgoing } : null,
        props: [...this.props.values()]
          .map((prop) => ({ ...prop }))
          .sort((a, b) => a.zIndex - b.zIndex || a.id.localeCompare(b.id)),
        dialogueVisible: this.dialogueVisible,
        dialogueCharacterId: this.dialogueCharacterId,
        dialogueWidthPercent: this.dialogueWidthPercent,
        loadedAssetKeys: this.assets.loadedKeys().sort(),
      };
    }
    return this.cachedSnapshot;
  }

  /** Props that currently exist in the scene, visible or not. */
  getProp(id: string): ScenePropState | undefined {
    const prop = this.props.get(id);
    return prop ? { ...prop } : undefined;
  }

  isAssetLoaded(assetId: string, variationId?: string): boolean {
    return this.assets.isLoaded(assetReferenceKey(assetId, variationId));
  }

  isDialogueVisible(): boolean {
    return this.dialogueVisible;
  }

  private cancelBackgroundTween(): void {
    this.backgroundTweenAbort?.abort();
    this.backgroundTweenAbort = null;
  }

  /** Abort an in-flight background fade and snap it to the end state. */
  private settleBackgroundTransition(): void {
    this.cancelBackgroundTween();
    if (this.outgoing) this.finishBackgroundTransition();
  }

  private runBackgroundTween(
    durationMs: number,
    onUpdate: (progress: number) => void,
    signal?: AbortSignal
  ): Promise<void> {
    const controller = new AbortController();
    this.backgroundTweenAbort = controller;
    const onAbort = () => controller.abort();
    if (signal?.aborted) controller.abort();
    else signal?.addEventListener("abort", onAbort, { once: true });

    const run = async () => {
      try {
        await this.tween(durationMs, onUpdate, controller.signal);
        if (this.backgroundTweenAbort === controller) {
          this.backgroundTweenAbort = null;
          this.finishBackgroundTransition();
        }
      } catch (error) {
        if (error instanceof DOMException && error.name === "AbortError") return;
        throw error;
      } finally {
        signal?.removeEventListener("abort", onAbort);
      }
    };

    if (this.skipTransitions) return run();
    void run();
    return Promise.resolve();
  }

  private notify(): void {
    this.cachedSnapshot = null;
    for (const listener of this.listeners) listener();
  }

  private tween(
    durationMs: number,
    onUpdate: (progress: number) => void,
    signal?: AbortSignal
  ): Promise<void> {
    return runTween({
      durationMs: this.skipTransitions ? 0 : durationMs,
      easing: this.easing,
      scheduler: this.scheduler,
      signal,
      onUpdate: (progress) => {
        onUpdate(progress);
        this.notify();
      },
    });
  }

  private requireProp(id: string, verb: string): ScenePropState {
    const prop = this.props.get(id);
    if (!prop) throw missingProp(id, verb);
    return prop;
  }

  private makeBackground(assetId: string, opacity: number, offset: StageVector): SceneBackgroundState {
    const assetKey = assetReferenceKey(assetId);
    this.assets.acquire(assetKey);
    return { assetId, assetKey, opacity, offsetX: offset.x, offsetY: offset.y };
  }

  private releaseBackground(background: SceneBackgroundState | null): void {
    if (background) this.assets.release(background.assetKey);
  }

  private beginBackgroundTransition(assetId: string, offset: StageVector, opacity: number): void {
    this.outgoing = this.background;
    this.background = this.makeBackground(assetId, opacity, offset);
  }

  private finishBackgroundTransition(): void {
    this.releaseBackground(this.outgoing);
    this.outgoing = null;
    if (this.background) {
      this.background.opacity = 1;
      this.background.offsetX = 0;
      this.background.offsetY = 0;
    }
    this.notify();
  }

  /** Offscreen slide offset in stage units for a direction the asset enters from. */
  private slideOffset(direction: Direction): StageVector {
    const from = offstagePosition({ x: 0, y: 0 }, direction);
    return {
      x: direction === "Top" || direction === "Bottom" ? 0 : from.x > 0 ? STAGE_WIDTH : -STAGE_WIDTH,
      y:
        direction === "Left" || direction === "Right"
          ? 0
          : from.y > 0
            ? STAGE_HEIGHT
            : -STAGE_HEIGHT,
    };
  }

  async applyOp(op: SceneOp, signal?: AbortSignal): Promise<void> {
    switch (op.kind) {
      case "bg.show": {
        this.settleBackgroundTransition();
        const previous = this.background;
        this.background = this.makeBackground(op.assetId, 1, { x: 0, y: 0 });
        this.releaseBackground(previous);
        this.notify();
        return;
      }

      case "bg.clear": {
        this.settleBackgroundTransition();
        this.releaseBackground(this.background);
        this.releaseBackground(this.outgoing);
        this.background = null;
        this.outgoing = null;
        this.notify();
        return;
      }

      case "bg.fade": {
        this.settleBackgroundTransition();
        this.beginBackgroundTransition(op.assetId, { x: 0, y: 0 }, 0);
        const incoming = this.background;
        const outgoing = this.outgoing;
        this.notify();
        await this.runBackgroundTween(
          op.durationMs,
          (progress) => {
            if (incoming) incoming.opacity = progress;
            if (outgoing) outgoing.opacity = 1 - progress;
          },
          signal
        );
        return;
      }

      case "bg.slideIn": {
        this.settleBackgroundTransition();
        const offset = this.slideOffset(op.direction);
        this.beginBackgroundTransition(op.assetId, offset, 1);
        const incoming = this.background;
        this.notify();
        await this.tween(
          op.durationMs,
          (progress) => {
            if (!incoming) return;
            incoming.offsetX = lerp(offset.x, 0, progress);
            incoming.offsetY = lerp(offset.y, 0, progress);
          },
          signal
        );
        this.finishBackgroundTransition();
        return;
      }

      case "bg.slideOut": {
        this.settleBackgroundTransition();
        const current = this.background;
        if (!current) return;
        const offset = this.slideOffset(op.direction);
        await this.tween(
          op.durationMs,
          (progress) => {
            current.offsetX = lerp(0, -offset.x, progress);
            current.offsetY = lerp(0, -offset.y, progress);
          },
          signal
        );
        this.releaseBackground(current);
        if (this.background === current) this.background = null;
        this.notify();
        return;
      }

      case "prop.add": {
        if (this.props.has(op.id)) {
          throw new SceneActionError(
            `Cannot add prop '${op.id}': a prop with this ID already exists in the scene. Use a different instance id or call Prop.Remove first.`
          );
        }
        const assetKey = assetReferenceKey(op.assetId, op.variationId);
        this.assets.acquire(assetKey);
        const position: StagePosition = { kind: "slot", slot: "Centre" };
        const coords = resolvePosition(position);
        this.props.set(op.id, {
          id: op.id,
          assetId: op.assetId,
          variationId: op.variationId,
          assetKey,
          position,
          x: coords.x,
          y: coords.y,
          visible: false,
          opacity: 1,
          zIndex: DEFAULT_PROP_Z,
          scale: 1,
          highlighted: false,
        });
        this.notify();
        return;
      }

      case "prop.remove": {
        const prop = this.requireProp(op.id, "remove");
        this.assets.release(prop.assetKey);
        this.props.delete(op.id);
        this.notify();
        return;
      }

      case "prop.show": {
        const prop = this.requireProp(op.id, "show");
        if (op.position) this.setPropPosition(prop, op.position);
        prop.visible = true;
        prop.opacity = 1;
        this.notify();
        return;
      }

      case "prop.hide": {
        const prop = this.requireProp(op.id, "hide");
        prop.visible = false;
        this.notify();
        return;
      }

      case "prop.fadeIn": {
        const prop = this.requireProp(op.id, "fade in");
        if (op.position) this.setPropPosition(prop, op.position);
        prop.visible = true;
        prop.opacity = 0;
        this.notify();
        await this.tween(op.durationMs, (progress) => {
          prop.opacity = progress;
        }, signal);
        prop.opacity = 1;
        this.notify();
        return;
      }

      case "prop.fadeOut": {
        const prop = this.requireProp(op.id, "fade out");
        const from = prop.opacity;
        await this.tween(op.durationMs, (progress) => {
          prop.opacity = lerp(from, 0, progress);
        }, signal);
        prop.opacity = 0;
        prop.visible = false;
        this.notify();
        return;
      }

      case "prop.slideIn": {
        const prop = this.requireProp(op.id, "slide in");
        const target = resolvePosition(op.position);
        const start = offstagePosition(target, op.direction);
        prop.position = op.position;
        prop.x = start.x;
        prop.y = start.y;
        prop.visible = true;
        prop.opacity = 1;
        this.notify();
        await this.tween(op.durationMs, (progress) => {
          prop.x = lerp(start.x, target.x, progress);
          prop.y = lerp(start.y, target.y, progress);
        }, signal);
        prop.x = target.x;
        prop.y = target.y;
        this.notify();
        return;
      }

      case "prop.slideOut": {
        const prop = this.requireProp(op.id, "slide out");
        const start = { x: prop.x, y: prop.y };
        const target = offstagePosition(start, op.direction);
        await this.tween(op.durationMs, (progress) => {
          prop.x = lerp(start.x, target.x, progress);
          prop.y = lerp(start.y, target.y, progress);
        }, signal);
        prop.visible = false;
        // Restore the authored position so a later Show does not appear offscreen.
        const restored = resolvePosition(prop.position);
        prop.x = restored.x;
        prop.y = restored.y;
        this.notify();
        return;
      }

      case "prop.move": {
        const prop = this.requireProp(op.id, "move");
        const start = { x: prop.x, y: prop.y };
        const target = resolvePosition(op.position);
        await this.tween(op.durationMs, (progress) => {
          prop.x = lerp(start.x, target.x, progress);
          prop.y = lerp(start.y, target.y, progress);
        }, signal);
        prop.position = op.position;
        prop.x = target.x;
        prop.y = target.y;
        this.notify();
        return;
      }

      case "prop.setPosition": {
        const prop = this.requireProp(op.id, "position");
        this.setPropPosition(prop, op.position);
        this.notify();
        return;
      }

      case "prop.setZ": {
        const prop = this.requireProp(op.id, "set the z layer of");
        prop.zIndex = Math.trunc(op.z);
        this.notify();
        return;
      }

      case "prop.setVariation": {
        const prop = this.requireProp(op.id, "set the variation of");
        const nextKey = assetReferenceKey(prop.assetId, op.variationId);
        if (nextKey === prop.assetKey) return;
        this.assets.acquire(nextKey);
        this.assets.release(prop.assetKey);
        prop.assetKey = nextKey;
        prop.variationId = op.variationId;
        this.notify();
        return;
      }

      case "prop.highlight": {
        const prop = this.requireProp(op.id, "highlight");
        prop.highlighted = true;
        this.notify();
        return;
      }

      case "prop.unhighlight": {
        const prop = this.requireProp(op.id, "remove the highlight from");
        prop.highlighted = false;
        this.notify();
        return;
      }

      case "dialogue.show": {
        this.dialogueVisible = true;
        this.dialogueCharacterId = op.characterId;
        this.notify();
        return;
      }

      case "dialogue.hide": {
        this.dialogueVisible = false;
        this.notify();
        return;
      }

      case "dialogue.setWidth": {
        if (
          !Number.isInteger(op.widthPercent) ||
          op.widthPercent < 1 ||
          op.widthPercent > 100
        ) {
          throw new SceneActionError(
            `Dialogue width must be an integer from 1 to 100 percent, got ${op.widthPercent}.`
          );
        }
        this.dialogueWidthPercent = op.widthPercent;
        this.notify();
        return;
      }

      default: {
        const exhaustive: never = op;
        throw new SceneActionError(`Unsupported scene operation: ${JSON.stringify(exhaustive)}`);
      }
    }
  }

  private setPropPosition(prop: ScenePropState, position: StagePosition): void {
    const coords = resolvePosition(position);
    prop.position = position;
    prop.x = coords.x;
    prop.y = coords.y;
  }

  /**
   * Dialogue boundary cleanup: any prop still hidden is destroyed and its asset
   * released, so scripts cannot rely on hidden props surviving between lines.
   */
  dialogueBoundary(): string[] {
    const removed: string[] = [];
    for (const [id, prop] of [...this.props.entries()]) {
      if (prop.visible) continue;
      this.assets.release(prop.assetKey);
      this.props.delete(id);
      removed.push(id);
    }
    if (removed.length > 0) this.notify();
    return removed;
  }

  /** Drop the whole stage, releasing every asset reference. */
  reset(): void {
    this.cancelBackgroundTween();
    this.props.clear();
    this.background = null;
    this.outgoing = null;
    this.dialogueVisible = true;
    this.dialogueCharacterId = undefined;
    this.dialogueWidthPercent = DEFAULT_DIALOGUE_WIDTH_PERCENT;
    this.assets.releaseAll();
    this.notify();
  }
}

File: scaffolds/unity/MuseLab/Assets/MuseLab/Scripts/Scene/SceneDirector.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace MuseLab.Scene
{
    public class ScenePropState
    {
        public string Id;
        public string AssetId;
        public string VariationId;
        public string AssetKey;
        public StagePosition Position;
        public double X;
        public double Y;
        public bool Visible;
        public float Opacity = 1f;
        public int ZIndex = 100;
        public float Scale = 1f;
        public bool Highlighted;
    }

    public class SceneBackgroundState
    {
        public string AssetId;
        public string AssetKey;
        public float Opacity = 1f;
        public double OffsetX;
        public double OffsetY;
    }

    public class SceneSnapshot
    {
        public SceneBackgroundState Background;
        public SceneBackgroundState OutgoingBackground;
        public List<ScenePropState> Props = new();
        public bool DialogueVisible = true;
        public string DialogueCharacterId;
        public int DialogueWidthPercent = SceneDirector.DefaultDialogueWidthPercent;
        public List<string> LoadedAssetKeys = new();
    }

    public class SceneActionException : Exception
    {
        public SceneActionException(string message) : base(message) { }
    }

    /// <summary>
    /// Authoritative Unity scene state. Visual actions resolve here rather than
    /// poking renderer objects directly.
    /// </summary>
    public class SceneDirector
    {
        public const int DefaultPropZ = 100;
        public const float HighlightScale = 1.08f;
        public const int DefaultDialogueWidthPercent = 50;

        readonly Dictionary<string, ScenePropState> props = new();
        readonly AssetReferenceCounter assets;
        readonly MonoBehaviour host;
        readonly bool skipTransitions;

        SceneBackgroundState background;
        SceneBackgroundState outgoing;
        bool dialogueVisible = true;
        string dialogueCharacterId;
        int dialogueWidthPercent = DefaultDialogueWidthPercent;
        Coroutine backgroundTween;
        int backgroundTweenGeneration;

        public event Action Changed;

        public SceneDirector(
            MonoBehaviour host,
            bool skipTransitions = false,
            Action<string> onLoadAsset = null,
            Action<string> onUnloadAsset = null)
        {
            this.host = host;
            this.skipTransitions = skipTransitions;
            assets = new AssetReferenceCounter(onLoadAsset, onUnloadAsset);
        }

        public SceneSnapshot GetSnapshot()
        {
            var snapshot = new SceneSnapshot
            {
                Background = CloneBackground(background),
                OutgoingBackground = CloneBackground(outgoing),
                LoadedAssetKeys = assets.LoadedKeys(),
            };
            foreach (var prop in props.Values)
                snapshot.Props.Add(CloneProp(prop));
            snapshot.Props.Sort((a, b) =>
            {
                var z = a.ZIndex.CompareTo(b.ZIndex);
                return z != 0 ? z : string.CompareOrdinal(a.Id, b.Id);
            });
            snapshot.DialogueVisible = dialogueVisible;
            snapshot.DialogueCharacterId = dialogueCharacterId;
            snapshot.DialogueWidthPercent = dialogueWidthPercent;
            return snapshot;
        }

        public IEnumerator ApplyOp(SceneOp op)
        {
            switch (op.Kind)
            {
                case SceneOpKind.BgShow:
                {
                    SettleBackgroundTransition();
                    var previous = background;
                    background = MakeBackground(op.AssetId, 1f, 0, 0);
                    ReleaseBackground(previous);
                    Notify();
                    yield break;
                }
                case SceneOpKind.BgClear:
                    SettleBackgroundTransition();
                    ReleaseBackground(background);
                    ReleaseBackground(outgoing);
                    background = null;
                    outgoing = null;
                    Notify();
                    yield break;
                case SceneOpKind.BgFade:
                {
                    SettleBackgroundTransition();
                    BeginBackgroundTransition(op.AssetId, 0, 0, 0f);
                    var incoming = background;
                    var outgoingBg = outgoing;
                    Notify();
                    if (skipTransitions)
                    {
                        yield return Tween(op.DurationMs, progress =>
                        {
                            if (incoming != null) incoming.Opacity = progress;
                            if (outgoingBg != null) outgoingBg.Opacity = 1f - progress;
                        });
                        FinishBackgroundTransition();
                        yield break;
                    }

                    var generation = backgroundTweenGeneration;
                    backgroundTween = host.StartCoroutine(RunBackgroundFade(incoming, outgoingBg, op.DurationMs, generation));
                    yield break;
                }
                case SceneOpKind.BgSlideIn:
                {
                    SettleBackgroundTransition();
                    var offset = SlideOffset(op.Direction);
                    BeginBackgroundTransition(op.AssetId, offset.x, offset.y, 1f);
                    var incoming = background;
                    Notify();
                    yield return Tween(op.DurationMs, progress =>
                    {
                        if (incoming == null) return;
                        incoming.OffsetX = SceneTransitions.Lerp((float)offset.x, 0f, progress);
                        incoming.OffsetY = SceneTransitions.Lerp((float)offset.y, 0f, progress);
                    });
                    FinishBackgroundTransition();
                    yield break;
                }
                case SceneOpKind.BgSlideOut:
                {
                    SettleBackgroundTransition();
                    var current = background;
                    if (current == null) yield break;
                    var offset = SlideOffset(op.Direction);
                    yield return Tween(op.DurationMs, progress =>
                    {
                        current.OffsetX = SceneTransitions.Lerp(0f, (float)-offset.x, progress);
                        current.OffsetY = SceneTransitions.Lerp(0f, (float)-offset.y, progress);
                    });
                    ReleaseBackground(current);
                    if (ReferenceEquals(background, current)) background = null;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropAdd:
                {
                    if (props.ContainsKey(op.Id))
                        throw new SceneActionException(
                            $"Cannot add prop '{op.Id}': a prop with this ID already exists in the scene.");
                    var assetKey = AssetReferenceCounter.Key(op.AssetId, op.VariationId);
                    assets.Acquire(assetKey);
                    var position = StagePosition.FromSlot("Centre");
                    var coords = StageCoords.Resolve(position);
                    props[op.Id] = new ScenePropState
                    {
                        Id = op.Id,
                        AssetId = op.AssetId,
                        VariationId = op.VariationId,
                        AssetKey = assetKey,
                        Position = position,
                        X = coords.x,
                        Y = coords.y,
                        Visible = false,
                        Opacity = 1f,
                        ZIndex = DefaultPropZ,
                        Scale = 1f,
                        Highlighted = false,
                    };
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropRemove:
                {
                    var prop = RequireProp(op.Id, "remove");
                    assets.Release(prop.AssetKey);
                    props.Remove(op.Id);
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropShow:
                {
                    var prop = RequireProp(op.Id, "show");
                    if (op.HasPosition) SetPropPosition(prop, op.Position);
                    prop.Visible = true;
                    prop.Opacity = 1f;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropHide:
                {
                    var prop = RequireProp(op.Id, "hide");
                    prop.Visible = false;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropFadeIn:
                {
                    var prop = RequireProp(op.Id, "fade in");
                    if (op.HasPosition) SetPropPosition(prop, op.Position);
                    prop.Visible = true;
                    prop.Opacity = 0f;
                    Notify();
                    yield return Tween(op.DurationMs, progress => prop.Opacity = progress);
                    prop.Opacity = 1f;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropFadeOut:
                {
                    var prop = RequireProp(op.Id, "fade out");
                    var from = prop.Opacity;
                    yield return Tween(op.DurationMs, progress =>
                        prop.Opacity = SceneTransitions.Lerp(from, 0f, progress));
                    prop.Opacity = 0f;
                    prop.Visible = false;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropSlideIn:
                {
                    var prop = RequireProp(op.Id, "slide in");
                    var target = StageCoords.Resolve(op.Position);
                    var start = StageCoords.Offstage(target, op.Direction);
                    prop.Position = op.Position;
                    prop.X = start.x;
                    prop.Y = start.y;
                    prop.Visible = true;
                    prop.Opacity = 1f;
                    Notify();
                    yield return Tween(op.DurationMs, progress =>
                    {
                        prop.X = SceneTransitions.Lerp((float)start.x, (float)target.x, progress);
                        prop.Y = SceneTransitions.Lerp((float)start.y, (float)target.y, progress);
                    });
                    prop.X = target.x;
                    prop.Y = target.y;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropSlideOut:
                {
                    var prop = RequireProp(op.Id, "slide out");
                    var start = (prop.X, prop.Y);
                    var target = StageCoords.Offstage(start, op.Direction);
                    yield return Tween(op.DurationMs, progress =>
                    {
                        prop.X = SceneTransitions.Lerp((float)start.X, (float)target.x, progress);
                        prop.Y = SceneTransitions.Lerp((float)start.Y, (float)target.y, progress);
                    });
                    prop.Visible = false;
                    var restored = StageCoords.Resolve(prop.Position);
                    prop.X = restored.x;
                    prop.Y = restored.y;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropMove:
                {
                    var prop = RequireProp(op.Id, "move");
                    var start = (prop.X, prop.Y);
                    var target = StageCoords.Resolve(op.Position);
                    yield return Tween(op.DurationMs, progress =>
                    {
                        prop.X = SceneTransitions.Lerp((float)start.X, (float)target.x, progress);
                        prop.Y = SceneTransitions.Lerp((float)start.Y, (float)target.y, progress);
                    });
                    prop.Position = op.Position;
                    prop.X = target.x;
                    prop.Y = target.y;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropSetPosition:
                {
                    var prop = RequireProp(op.Id, "position");
                    SetPropPosition(prop, op.Position);
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropSetZ:
                {
                    var prop = RequireProp(op.Id, "set the z layer of");
                    prop.ZIndex = op.Z;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropSetVariation:
                {
                    var prop = RequireProp(op.Id, "set the variation of");
                    var nextKey = AssetReferenceCounter.Key(prop.AssetId, op.VariationId);
                    if (nextKey == prop.AssetKey) yield break;
                    assets.Acquire(nextKey);
                    assets.Release(prop.AssetKey);
                    prop.AssetKey = nextKey;
                    prop.VariationId = op.VariationId;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropHighlight:
                {
                    var prop = RequireProp(op.Id, "highlight");
                    prop.Highlighted = true;
                    Notify();
                    yield break;
                }
                case SceneOpKind.PropUnhighlight:
                {
                    var prop = RequireProp(op.Id, "remove the highlight from");
                    prop.Highlighted = false;
                    Notify();
                    yield break;
                }
                case SceneOpKind.DialogueShow:
                    dialogueVisible = true;
                    dialogueCharacterId = string.IsNullOrEmpty(op.CharacterId) ? null : op.CharacterId;
                    Notify();
                    yield break;
                case SceneOpKind.DialogueHide:
                    dialogueVisible = false;
                    Notify();
                    yield break;
                case SceneOpKind.DialogueSetWidth:
                    if (op.WidthPercent < 1 || op.WidthPercent > 100)
                        throw new SceneActionException(
                            $"Dialogue width must be an integer from 1 to 100 percent, got {op.WidthPercent}.");
                    dialogueWidthPercent = op.WidthPercent;
                    Notify();
                    yield break;
                default:
                    throw new SceneActionException($"Unsupported scene operation: {op.Kind}");
            }
        }

        public List<string> DialogueBoundary()
        {
            var removed = new List<string>();
            foreach (var entry in new List<KeyValuePair<string, ScenePropState>>(props))
            {
                if (entry.Value.Visible) continue;
                assets.Release(entry.Value.AssetKey);
                props.Remove(entry.Key);
                removed.Add(entry.Key);
            }
            if (removed.Count > 0) Notify();
            return removed;
        }

        public void Reset()
        {
            CancelBackgroundTween();
            props.Clear();
            background = null;
            outgoing = null;
            dialogueVisible = true;
            dialogueCharacterId = null;
            dialogueWidthPercent = DefaultDialogueWidthPercent;
            assets.ReleaseAll();
            Notify();
        }

        IEnumerator RunBackgroundFade(
            SceneBackgroundState incoming,
            SceneBackgroundState outgoingBg,
            int durationMs,
            int generation)
        {
            yield return Tween(durationMs, progress =>
            {
                if (incoming != null) incoming.Opacity = progress;
                if (outgoingBg != null) outgoingBg.Opacity = 1f - progress;
            });
            if (generation != backgroundTweenGeneration) yield break;
            backgroundTween = null;
            FinishBackgroundTransition();
        }

        void CancelBackgroundTween()
        {
            backgroundTweenGeneration++;
            if (backgroundTween == null) return;
            host.StopCoroutine(backgroundTween);
            backgroundTween = null;
        }

        void SettleBackgroundTransition()
        {
            CancelBackgroundTween();
            if (outgoing != null) FinishBackgroundTransition();
        }

        IEnumerator Tween(int durationMs, Action<float> onUpdate)
        {
            var duration = skipTransitions ? 0 : durationMs;
            yield return SceneTransitions.RunTween(duration, progress =>
            {
                onUpdate(progress);
                Notify();
            });
        }

        ScenePropState RequireProp(string id, string verb)
        {
            if (!props.TryGetValue(id, out var prop))
                throw new SceneActionException(
                    $"Cannot {verb} prop '{id}': no active prop with this ID exists. Call Prop.Add(...) first.");
            return prop;
        }

        void SetPropPosition(ScenePropState prop, StagePosition position)
        {
            var coords = StageCoords.Resolve(position);
            prop.Position = position;
            prop.X = coords.x;
            prop.Y = coords.y;
        }

        SceneBackgroundState MakeBackground(string assetId, float opacity, double offsetX, double offsetY)
        {
            var assetKey = AssetReferenceCounter.Key(assetId);
            assets.Acquire(assetKey);
            return new SceneBackgroundState
            {
                AssetId = assetId,
                AssetKey = assetKey,
                Opacity = opacity,
                OffsetX = offsetX,
                OffsetY = offsetY,
            };
        }

        void ReleaseBackground(SceneBackgroundState state)
        {
            if (state != null) assets.Release(state.AssetKey);
        }

        void BeginBackgroundTransition(string assetId, double offsetX, double offsetY, float opacity)
        {
            outgoing = background;
            background = MakeBackground(assetId, opacity, offsetX, offsetY);
        }

        void FinishBackgroundTransition()
        {
            ReleaseBackground(outgoing);
            outgoing = null;
            if (background != null)
            {
                background.Opacity = 1f;
                background.OffsetX = 0;
                background.OffsetY = 0;
            }
            Notify();
        }

        (double x, double y) SlideOffset(string direction)
        {
            var from = StageCoords.Offstage((0, 0), direction);
            return (
                direction is "Top" or "Bottom" ? 0 : from.x > 0 ? StageCoords.Width : -StageCoords.Width,
                direction is "Left" or "Right" ? 0 : from.y > 0 ? StageCoords.Height : -StageCoords.Height
            );
        }

        void Notify() => Changed?.Invoke();

        static SceneBackgroundState CloneBackground(SceneBackgroundState source) =>
            source == null
                ? null
                : new SceneBackgroundState
                {
                    AssetId = source.AssetId,
                    AssetKey = source.AssetKey,
                    Opacity = source.Opacity,
                    OffsetX = source.OffsetX,
                    OffsetY = source.OffsetY,
                };

        static ScenePropState CloneProp(ScenePropState source) =>
            new()
            {
                Id = source.Id,
                AssetId = source.AssetId,
                VariationId = source.VariationId,
                AssetKey = source.AssetKey,
                Position = source.Position,
                X = source.X,
                Y = source.Y,
                Visible = source.Visible,
                Opacity = source.Opacity,
                ZIndex = source.ZIndex,
                Scale = source.Scale,
                Highlighted = source.Highlighted,
            };
    }
}
